arrows/klv/klv_0601.cxx

```diff
@@ -3,430 +3,1432 @@
+  return klv_ber_oid_length( value.id ) +
+         klv_ber_length( klv_string_length( value.string ) ) +
+         klv_string_length( value.string ) +
+         ( value.timestamp ? 8 : 0 );
+}
 
-/** Globally available 0601 key.
- *
- */
-klv_uds_key
-klv_0601_key()
+// ----------------------------------------------------------------------------
+std::string
+klv_0601_control_command_format
+::description() const
 {
-  return klv_0601_uds_key;
+  return "control command of " + length_description();
 }
 
-bool
-is_klv_0601_key( klv_uds_key const& key )
+// ----------------------------------------------------------------------------
+std::ostream&
+operator<<( std::ostream& os, klv_0601_frame_rate const& value )
 {
-  return key == klv_0601_uds_key;
+  if( value.denominator == 1 )
+  {
+    return os << value.numerator;
+  }
+  if( value.denominator == 0 ) {
+    return os << "(invalid)";
+  }
+  return os << std::fixed << std::setprecision( 3 )
+            << ( static_cast< double >( value.numerator ) /
+                 value.denominator );
 }
 
-klv_0601_tag
-klv_0601_get_tag( klv_lds_key key )
+// ----------------------------------------------------------------------------
+bool
+operator==( klv_0601_frame_rate const& lhs, klv_0601_frame_rate const& rhs )
 {
-  return static_cast< klv_0601_tag > ( uint8_t( key ) );
+  return lhs.numerator == rhs.numerator && lhs.denominator == rhs.denominator;
 }
 
-// ----------------------------------------------------------------
-/** Compute 0601 checksum.
- *
- * Verify the 0601 block checksum against the expected checksum.  The
- * supplied data is a klv UDS packet which may contain multiple 0601
- * LDS data packets.
- *
- * The checksum is a running 16-bit sum through the entire UDS packet
- * starting with the 16 byte Local Data Set key and ending with
- * summing the length field of the checksum data item.
- *
- * Packet layout
- *
- *  |__| ... |__|__|__|__|
- *            ^  ^  ^--^----- 16 bit checksum
- *            |  |
- *            |   ----------- length of checksum in bytes (= 2)
- *             -------------- checksum packet type code (= 1)
- *
- * @param[in] data raw data to checksum
- *
- * @return \b True if checksum matches expected; \b False otherwise.
- * \b False is also returned if checksum tag is not found.
- */
+// ----------------------------------------------------------------------------
 bool
-klv_0601_checksum( klv_data const& data )
+operator<( klv_0601_frame_rate const& lhs, klv_0601_frame_rate const& rhs )
 {
-  klv_data::const_iterator_t eit = data.klv_end();
-
-  // if checksum tag is not where expected then terminate early
-  if ( ( *( eit - 4 ) != 0x01 ) && //
-       ( *( eit - 3 ) != 0x02 ) ) // cksum length
+  if( lhs.numerator < rhs.numerator )
   {
-    return false;
+    return true;
+  }
+  if( lhs.numerator == rhs.numerator )
+  {
+    return lhs.denominator < rhs.denominator;
   }
+  return false;
+}
```

Comment:
`lhs.numerator < rhs.numberator` only works if `lhs.denominator` == `rhs.denominator`.  But according to line 1252, the denominators are not necessarily equal.  Daniel knows the standard better than I and there may be a reason this can work, but it seems incomplete.

Reply:
This was because I wasn't really thinking of the frame rate as representing an actual number, per se, but as a data structure which needs some arbitrary but consistent ordering mechanism (provided by the `<` operator) for sorting purposes. You can see similar operators implemented for all the other kinds of objects ('klv_0601_control_command') where one being literally 'less than' another is nonsensical. I see your point where it is counterintuitive, and a more literal implementation would probably be better. I might wait for a future PR to change this code, though, since this PR is only copying-pasting these lines, not originating them.